File: p228e/summary_ranges.py
from typing import List


class Solution:
    def summaryRanges(self, nums: List[int]) -> List[str]:
        ranges = []
        for n in nums:
            if ranges and n == ranges[-1][1] + 1:
                ranges[-1][1] = n
            else:
                ranges.append([n, n])
        return [f"{l}->{r}" if l != r else str(l) for l, r in ranges]


# TESTS
for nums, expected in [
    ([0, 1, 2, 4, 5, 7], ["0->2", "4->5", "7"]),
    ([0, 2, 3, 4, 6, 8, 9], ["0", "2->4", "6", "8->9"]),
    ([], []),
    ([-1], ["-1"]),
    ([0], ["0"]),
]:
    sol = Solution()
    actual = sol.summaryRanges(nums)
    print("Summary ranges of", nums, "->", actual)
    assert actual == expected

File: p228e/summary_ranges.js
/**
 * @param {number[]} nums
 * @return {string[]}
 */
var summaryRanges = function (nums) {
  const ans = [];
  let start = nums[0];
  for (let i = 1; i <= nums.length; ++i) {
    if (nums[i] != nums[i - 1] + 1) {
      ans.push(start === nums[i - 1] ? `${start}` : `${start}->${nums[i - 1]}`);
      start = nums[i];
    }
  }
  return ans;
};
// TEST
[
  [0, 1, 2, 4, 5, 7],
  [0, 2, 3, 4, 6, 8, 9],
  [0, 2, 4, 6, 8, 10],
  [0, 1, 2, 3, 4, 5],
].forEach((nums) => {
  console.log('Summary ranges of', nums, '->', summaryRanges(nums));
});
